src/common/walFilter/versions/recordProcessGPDB6.c

```diff
@@ -0,0 +1,620 @@
+            *node = (RelFileNode *) XLogRecGetData(record);
+            return true;
+        }
+
+        default:
+            THROW_FMT(FormatError, "Appendonly UNKNOWN: %d", info);
+    }
+}
+
+FN_EXTERN bool
+getRelFileNodeGPDB6(const XLogRecord *record, RelFileNode **node)
+{
+    switch (record->xl_rmid)
+    {
+        case ResourceManager_XLOG:
+            return getXlog(record, node);
+
+        case ResourceManager_Storage:
+            return getStorage(record, node);
+
+        case ResourceManager_Heap2:
+            return getHeap2(record, node);
+
+        case ResourceManager_Heap:
+            return getHeap(record, node);
+
+        case ResourceManager_Btree:
+            return getBtree(record, node);
+
+        case ResourceManager_Gin:
+            return getGin(record, node);
+
+        case ResourceManager_Gist:
+            return getGist(record, node);
+
+        case ResourceManager_Sequence:
+            return getSeq(record, node);
+
+        case ResourceManager_SPGist:
+            return getSpgist(record, node);
+
+        case ResourceManager_Bitmap:
+            return getBitmap(record, node);
+
+        case ResourceManager_Appendonly:
+            return getAppendonly(record, node);
+
+        case ResourceManager_Transaction:
+        case ResourceManager_CLOG:
+        case ResourceManager_Database:
+        case ResourceManager_Tablespace:
+        case ResourceManager_MultiXact:
+        case ResourceManager_RelMap:
+        case ResourceManager_Standby:
+        case ResourceManager_DistributedLog:
+            // skip
+            return false;
+
+        case ResourceManager_Hash:
+            THROW(FormatError, "Not supported in greenplum. shouldn't be here");
+
+        default:
+            THROW(FormatError, "Unknown resource manager");
+    }
+}
+
+FN_EXTERN void
+validXLogRecordHeaderGPDB6(const XLogRecord *const record)
+{
+    /*
+     * xl_len == 0 is bad data for everything except XLOG SWITCH, where it is
+     * required.
+     */
+    if (record->xl_rmid == ResourceManager_XLOG && record->xl_info == XLOG_SWITCH)
+    {
+        if (record->xl_len != 0)
+        {
+            THROW_FMT(FormatError, "invalid xlog switch record");
+        }
+    }
+    else if (record->xl_len == 0)
+    {
+        THROW_FMT(FormatError, "record with zero length");
+    }
+    if (record->xl_tot_len < SizeOfXLogRecord + record->xl_len ||
+        record->xl_tot_len > SizeOfXLogRecord + record->xl_len +
+        XLR_MAX_BKP_BLOCKS * (sizeof(BkpBlock) + BLCKSZ))
+    {
+        THROW_FMT(FormatError, "invalid record length");
+    }
+    if (record->xl_rmid > ResourceManager_Appendonly)
+    {
+        THROW_FMT(FormatError, "invalid resource manager ID %u", record->xl_rmid);
+    }
+}
+
+FN_EXTERN void
+validXLogRecordGPDB6(const XLogRecord *const record)
+{
+    pg_crc32 crc;
+    int i;
+    uint32_t len = record->xl_len;
+    BkpBlock bkpb;
+    char *blk;
+    size_t remaining = record->xl_tot_len;
```

Comment:
If we change the original function quite a bit, do we need declarations at the beginning of the block?